base/strings/util.jl

```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
 end
 
 """
+    hex2bytes(s::AbstractVector{UInt8})
+
+Convert the hexadecimal bytes array to its binary representation. Returns
+`Vector{UInt8}`, i.e. a vector of bytes.
+"""
+@inline function hex2bytes(s::AbstractVector{UInt8})
+    len = length(s)
+    if isodd(len)
+        throw(ArgumentError("source vector length must be even"))
+    end
+    d = zeros(UInt8, div(len, 2))
+    return hex2bytes!(d, s)
+end
+
+"""
+    hex2bytes!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
+
+Converts the hexadecimal bytes vector to its binary representation. The results are
+populated into a destination vector. The function returns the destination vector.
+
+# Examples
+```jldoctest
+julia> s = UInt8["01abEF"...]
+6-element Array{UInt8,1}:
+ 0x30
+ 0x31
+ 0x61
+ 0x62
+ 0x45
+ 0x46
+
+julia> d = zeros(UInt8, 3)
+3-element Array{UInt8,1}:
+ 0x00
+ 0x00
+ 0x00
+
+julia> hex2bytes!(d, s)
+3-element Array{UInt8,1}:
+ 0x01
+ 0xab
+ 0xef
+```
+"""
+function hex2bytes!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
+    i, j = start(s), 0
+    # This line is important as this ensures computation happens in word boundary and not
+    # byte boundary. Boundary computation can be almost 10 times slower
+    n::UInt = 0
+    c1::UInt = 0
+    c2::UInt = 0
+    while !done(s, i)
+        n = 0
+        c1, i = next(s, i)
+        done(s, i) && throw(ArgumentError("source vector length must be even"))
+        c2, i = next(s, i)
+        n = number_from_hex(c1)
+        n <<= 4
+        n += number_from_hex(c2)
+        d[j+=1] = (n & 0xFF)
+    end
+    return d
+end
+
+@inline function number_from_hex(c::UInt)
+    DIGIT_ZERO     = UInt('0')
+    DIGIT_NINE     = UInt('9')
+    LATIN_UPPER_A  = UInt('A')
+    LATIN_UPPER_F  = UInt('F')
+    LATIN_A        = UInt('a')
+    LATIN_F        = UInt('f')
```

Comment:
I don't think defining these constants adds much to the clarity (on the contrary, it makes the code longer and hence harder to read); it would be just as compact and readable if you used `UInt8('0')` etcetera directly in the code.